compiler/rustc_typeck/src/check/method/probe.rs

```diff
@@ -612,8 +612,55 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
     fn assemble_probe(&mut self, self_ty: &Canonical<'tcx, QueryResponse<'tcx, Ty<'tcx>>>) {
         debug!("assemble_probe: self_ty={:?}", self_ty);
         let lang_items = self.tcx.lang_items();
+        let error_def_id = self.tcx.require_lang_item(hir::LangItem::Error, None);
+        let send_def_id = self.tcx.require_lang_item(hir::LangItem::Send, None);
+        let sync_def_id = self.tcx.require_lang_item(hir::LangItem::Sync, None);
 
         match *self_ty.value.value.kind() {
+            ty::Dynamic(data, ..) if data.principal_def_id() == Some(error_def_id) => {
+                let (QueryResponse { value: generalized_self_ty, .. }, _ignored_var_values) =
+                    self.fcx
+                        .instantiate_canonical_with_fresh_inference_vars(self.span, self_ty);
+
+                self.assemble_inherent_candidates_from_object(generalized_self_ty);
+
+                let (mut is_send, mut is_sync) = (false, false);
+                let bounds = data.auto_traits();
+
+                self.assemble_inherent_impl_candidates_for_type(error_def_id);
+
+                for bound in bounds {
+                    if bound == send_def_id {
+                        is_send = true;
+                    } else if bound == sync_def_id {
+                        is_sync = true;
+                    } else {
+                        panic!("unexpected bound: {:?}", bound);
+                    }
+                }
+
+                match (is_send, is_sync) {
+                    (false, false) => {
+                        let lang_def_id = lang_items.error_impl();
+                        self.assemble_inherent_impl_for_primitive(lang_def_id);
+                        let lang_def_id = lang_items.error_alloc_impl();
+                        self.assemble_inherent_impl_for_primitive(lang_def_id);
+                    }
+                    (true, false) => {
+                        let lang_def_id = lang_items.errorsend_impl();
+                        self.assemble_inherent_impl_for_primitive(lang_def_id);
+                        let lang_def_id = lang_items.errorsend_alloc_impl();
+                        self.assemble_inherent_impl_for_primitive(lang_def_id);
+                    }
+                    (true, true) => {
+                        let lang_def_id = lang_items.errorsendsync_impl();
+                        self.assemble_inherent_impl_for_primitive(lang_def_id);
+                        let lang_def_id = lang_items.errorsendsync_alloc_impl();
+                        self.assemble_inherent_impl_for_primitive(lang_def_id);
+                    }
+                    (false, true) => panic!("unexpected combination of marker traits"),
+                }
+            }
```

Comment:
This bit is currently incorrect, I yolo'ed it to see what would happen and it was not good. Trying to get a better understanding of what `assemble_probe` actually does now before proceeding.